scheduler/preemption.go

```diff
@@ -0,0 +1,617 @@
+
+	var bestAllocs []*structs.Allocation
+	allRequirementsMet := false
+
+	// Initialize variable to track resources as they become available from preemption
+	availableResources := p.nodeRemainingResources.Copy()
+
+	// Iterate over allocations grouped by priority to find preemptible allocations
+	for _, allocGrp := range allocsByPriority {
+		for len(allocGrp.allocs) > 0 && !allRequirementsMet {
+			closestAllocIndex := -1
+			bestDistance := math.MaxFloat64
+			// Find the alloc with the closest distance
+			for index, alloc := range allocGrp.allocs {
+				currentPreemptionCount := p.getNumPreemptions(alloc)
+				allocDetails := p.allocDetails[alloc.ID]
+				maxParallel := allocDetails.maxParallel
+				distance := scoreForTaskGroup(resourcesNeeded, allocDetails.resources, maxParallel, currentPreemptionCount)
+				if distance < bestDistance {
+					bestDistance = distance
+					closestAllocIndex = index
+				}
+			}
+			closestAlloc := allocGrp.allocs[closestAllocIndex]
+			closestResources := p.allocDetails[closestAlloc.ID].resources
+			availableResources.Add(closestResources)
+
+			// This step needs the original resources asked for as the second arg, can't use the running total
+			allRequirementsMet, _ = availableResources.Superset(resourceAsk.Comparable())
+
+			bestAllocs = append(bestAllocs, closestAlloc)
+
+			allocGrp.allocs[closestAllocIndex] = allocGrp.allocs[len(allocGrp.allocs)-1]
+			allocGrp.allocs = allocGrp.allocs[:len(allocGrp.allocs)-1]
+
+			// This is the remaining total of resources needed
+			resourcesNeeded.Subtract(closestResources)
+		}
+		if allRequirementsMet {
+			break
+		}
+	}
+
+	// Early return if all allocs examined and requirements were not met
+	if !allRequirementsMet {
+		return nil
+	}
+
+	// We do another pass to eliminate unnecessary preemptions
+	// This filters out allocs whose resources are already covered by another alloc
+	basePreemptionResource := GetBasePreemptionResourceFactory()
+	resourcesNeeded = resourceAsk.Comparable()
+	filteredBestAllocs := p.filterSuperset(bestAllocs, p.nodeRemainingResources, resourcesNeeded, basePreemptionResource)
+	return filteredBestAllocs
+
+}
+
+// PreemptForNetwork tries to find allocations to preempt to meet network resources.
+// This is called once per task when assigning a network to the task. While finding allocations
+// to preempt, this only considers allocations that share the same network device
+func (p *Preemptor) PreemptForNetwork(networkResourceAsk *structs.NetworkResource, netIdx *structs.NetworkIndex) []*structs.Allocation {
+
+	// Early return if there are no current allocs
+	if len(p.currentAllocs) == 0 {
+		return nil
+	}
+
+	deviceToAllocs := make(map[string][]*structs.Allocation)
+	MbitsNeeded := networkResourceAsk.MBits
+	reservedPortsNeeded := networkResourceAsk.ReservedPorts
+
+	// Build map of reserved ports needed for fast access
+	reservedPorts := make(map[int]interface{})
```

Comment:
Use struct{} instead of interface{}. It is a smaller object and more canonical for set implementation.